manual/src/refman/extensions/tail_mod_cons.etex

```diff
@@ -0,0 +1,548 @@
+positions become tail positions; after this transformation, the
+implementation of "map" above becomes \emph{tail-recursive}, in the
+sense that it only consumes a constant amount of stack space. The
+OCaml compiler implements this transformation on demand, using the
+"[\@tail_mod_cons]" or "[\@ocaml.tail_mod_cons]" attribute on the
+function to transform.
+
+\begin{verbatim}
+let[@tail_mod_cons] rec map f li =
+  match li with
+  | [] -> []
+  | x :: xs ->
+    let y = f x in
+    y :: map f xs
+\end{verbatim}
+
+\begin{verbatim}
+# List.length (map Fun.id (List.init 1_000_000 (fun i -> i)));;
+- : int = 1000000
+\end{verbatim}
+
+This transformation only improves calls in tail-modulo-cons position,
+it does not improve recursive calls that do not fit in this fragment:
+\begin{verbatim}
+(* does *not* work: addition is not a data constructor *)
+  let[@tail_mod_cons] rec length li =
+    match li with
+    | [] -> 0
+    | _ :: xs -> 1 + length li
+
+Warning 71 [unused-tmc-attribute]: This function is marked @tail_mod_cons
+but is never applied in TMC position.
+\end{verbatim}
+
+It is of course possible to use the "[\@tail_mod_cons]" transformation
+on programs that contain some recursive calls in tail-modulo-cons
+position, and some calls in other, arbitrary positions. Only the tail
+calls and tail-modulo-cons calls will happen in constant stack space.
+
+\paragraph{General design} This feature is provided as an explicit
+program transformation, not an implicit optimization. It is
+annotation-driven: the user is expected to express their intent by
+adding annotations in the program, and will be asked to do so in any
+ambiguous situation.
+
+We expect it to be used mostly by advanced OCaml users needing to get
+some guarantees on the stack-consumption behavior of their
+programs. Our recommendation is to use the "[\@tailcall]" annotation on
+all callsites that should not consume any stack
+space. "[\@tail_mod_cons]" extends the set of programs on which calls
+can be annotated to be tail, helping establish stack-consumption
+guarantees in more cases.
+
+\paragraph{Performance} A standard approach to get a tail-recursive
+version of "List.map" is to use an accumulator to collect output
+elements, and reverse it at the end of the traversal.
+
+\begin{verbatim}
+let rec map f li = map_aux f [] li
+and map_aux f acc li =
+  match li with
+  | [] -> List.rev acc
+  | x :: xs ->
+    let y = f x in
+    map_aux f (y :: acc) xs
+\end{verbatim}
+
+This version is tail-recursive, but it is measurably slower than the
+simple, non-tail-recursive version. In contrast, the tail-mod-cons
+transformation provides an implementation that has comparable
+performance to the original version, even on small inputs.
+
+\paragraph{Evaluation order} Beware that the tail-modulo-cons
+transformation has an effect on evaluation order: the constructor
+argument that is transformed into tail-position will always be
+evaluated last. Consider the following example:
```

Comment:
I don't think `right-to-left` would be the correct evaluation order to specify if we wanted to specify one, obviously left-to-right is the choice that will not "surprise users".

My belief is that pure function calls are common enough that being forced to write programs in A-normal-forms to avoid effect interleaving would feel like a significant *and* unjust burden to users. I made the argument in https://github.com/ocaml/ocaml/pull/9760#issuecomment-727540945 , and my view hasn't really changed since.

If we were to find out that I was totally wrong and this some-argument-goes-last business is a big issue, we could always make things more restrictive in the future (I don't disagree with your point (2)), as users would then have to rewrite their code in a way that is backward-compatible.

Reply:
I checked once - almost everywhere the Compiler uses _arabic_ style of evaluation:

```ocaml

f arg3 arg2 arg1

[ arg3; arg2; arg1]

arg3::arg2::arg1::[]

( arg3, arg2, arg1 )

[| arg3; arg2; arg1 |]
```

Only 

```ocaml
{| arg3, arg2 arg1 |}
```

shows them in the reverse order. ;-)

Reply:
Thanks for the pointer to the place this was previously discussed. I think that part of my objection lifts if the annotation on function calls is made mandatory, because the control on evaluation order becomes local. Above I was merely suggesting to apply the advice to the documentation itself, but I also like your idea of making it enforced by the compiler. Obvious subsequent question: if all tail mod const calls were made explicit with an annotation, do you still need one on the function definition? (I did not find a discussion about this point.)

The discussion about changing evaluation order is fairly hypothetical, but I wanted to mention that there are more theoretical arguments in favour of right-to-left application as far as function applications are concerned (I can develop this point elsewhere if you are interested).